cpmpy/expressions/globalconstraints.py

```diff
@@ -272,6 +273,105 @@ def value(self):
         return pathlen == len(self.args) and idx == 0
 
 
+class SubCircuit(GlobalConstraint):
+    """
+        The sequence of variables form a subcircuits, where x[i] = j means that j is the successor of i.
+        Contrary to Circuit, there is no requirement on all nodes needing to be part of the circuit.
+        Nodes which aren't part of the subcircuit, should self loop i.e. x[i] = i.
+        The size of the subcircuit should be strictly greater than 1, so not all stops can selfloop.
+        When a startIndex is provided, it will be treated as the start of the subcircuit and will be garuanteed
+        to be inside the subcircuit.
+    """
+
+    def __init__(self, *args, startIndex:int=None):
+        flatargs = flatlist(args)
+        # Ensure all args are integer successor values
+        if any(is_boolexpr(arg) for arg in flatargs):
+            raise TypeError("SubCircuit global constraint only takes arithmetic arguments: {}".format(flatargs))
+        # Ensure startIndex is an integer
+        if not isinstance(startIndex, int) and startIndex is not None:
+            raise TypeError("SubCircuit global constraint's startIndex argument must be an integer: {}".format(startIndex))
+        super().__init__("subcircuit", flatargs)
+        self.startIndex = startIndex
+        # Ensure there are at least two stops to create a circuit with
```

Comment:
This one should go above the super() call, so all checks are done before creating any objects

Reply:
I just replicated the structure of Circuit (it also checks for the number of arguments after the super call). Should I fix that one too then?